File: 0207-course-schedule/0207-course-schedule.js
var canFinish = function(numCourses, prerequisites) {
  const helper = (v, z) => {
    if (visited[v]) {
      return false;
    }
    if (recStack[v]) {
      return true;
    }
    recStack[v] = true;
    const li = z.get(v) || [];
    for(let i = 0; i < li.length; i++) {
      if(helper(li[i], z)) {
        return true;
      }
    }
    visited[v] = true;
    recStack[v] = false;
    return false;
  }
  const visited = new Array(numCourses).fill(false);
  const recStack = new Array(numCourses).fill(false);
  const z = new Map();
  for(const [x, y] of prerequisites) {
    if(z.has(x)) {
      z.get(x).push(y);
    } else {
      z.set(x, [y]);
    }
  }
  for(let v = 0; v < numCourses; v++) {
    if (helper(v, z)) {
      return false;
    }
  }
  return true;
};

File: 0207-course-schedule/0207-course-schedule.go
func helper(v int, z map[int][]int, visited []bool, recStack []bool) bool {
	if visited[v] {
		return false
	}
	if recStack[v] {
		return true
	}
	recStack[v] = true
	if len(z[v]) > 0 {
		for _, i := range z[v] {
			if helper(i, z, visited, recStack) {
				return true
			}
		}
	}
	visited[v] = true
	recStack[v] = false // reset the recursion stack before exiting
	return false
}

func canFinish(numCourses int, prerequisites [][]int) bool {
	visited := make([]bool, numCourses)
	recStack := make([]bool, numCourses)
	z := make(map[int][]int)
	for i := 0; i < len(prerequisites); i++ {
		z[prerequisites[i][0]] = append(z[prerequisites[i][0]], prerequisites[i][1])
	}
	for v := 0; v < numCourses; v++ {
		if helper(v, z, visited, recStack) {
			return false
		}
	}
	return true
}